File: invoice_parse/libs/shared-py/invoice_shared/adapters/queue.py
"""Queue adapter — abstract interface + Redis Streams implementation."""

from __future__ import annotations

import json
import uuid
from abc import ABC, abstractmethod
from typing import Any, Callable

import redis


class MessageQueue(ABC):
    @abstractmethod
    def publish(self, topic: str, message: dict) -> str:
        """Publish a message. Returns message ID."""
        ...

    @abstractmethod
    def consume(self, topic: str, count: int = 1, block_ms: int = 5000) -> list[tuple[str, dict]]:
        """Consume messages. Returns list of (message_id, message_dict)."""
        ...

    @abstractmethod
    def ack(self, topic: str, message_id: str) -> None: ...

    @abstractmethod
    def extend_visibility(self, topic: str, message_id: str, seconds: int) -> None: ...


class RedisStreamQueue(MessageQueue):
    """Redis Streams with consumer groups.

    Uses XADD for publish, XREADGROUP for consume, XACK for acknowledge.
    Consumer groups ensure messages are only delivered to one consumer
    and can be reclaimed (XCLAIM) if a consumer crashes.
    """

    def __init__(self, url: str, consumer_group: str) -> None:
        self._client = redis.Redis.from_url(url, decode_responses=True)
        self._consumer_group = consumer_group
        self._consumer_name = f"worker-{uuid.uuid4().hex[:8]}"
        self._initialized_groups: set[str] = set()

    def _ensure_group(self, topic: str) -> None:
        if topic in self._initialized_groups:
            return
        try:
            self._client.xgroup_create(topic, self._consumer_group, id="0", mkstream=True)
        except redis.ResponseError as e:
            if "BUSYGROUP" not in str(e):
                raise
        self._initialized_groups.add(topic)

    def publish(self, topic: str, message: dict) -> str:
        self._ensure_group(topic)
        payload = {"data": json.dumps(message)}
        msg_id: str = self._client.xadd(topic, payload)
        return msg_id

    def consume(self, topic: str, count: int = 1, block_ms: int = 5000) -> list[tuple[str, dict]]:
        self._ensure_group(topic)
        results = self._client.xreadgroup(
            groupname=self._consumer_group,
            consumername=self._consumer_name,
            streams={topic: ">"},
            count=count,
            block=block_ms,
        )
        if not results:
            return []
        messages = []
        for _stream, entries in results:
            for msg_id, fields in entries:
                data = json.loads(fields["data"])
                messages.append((msg_id, data))
        return messages

    def ack(self, topic: str, message_id: str) -> None:
        self._client.xack(topic, self._consumer_group, message_id)

    def extend_visibility(self, topic: str, message_id: str, seconds: int) -> None:
        # Redis Streams don't have visibility timeout — no-op.
        # SQS implementation would call ChangeMessageVisibility here.
        pass

    def pending(self, topic: str) -> list[dict[str, Any]]:
        """Get pending (unacknowledged) messages — useful for monitoring and reaper."""
        return self._client.xpending_range(
            topic, self._consumer_group, min="-", max="+", count=100
        )

File: invoice-parse/libs/shared-rs/src/adapters/queue.rs
use redis::streams::{StreamReadOptions, StreamReadReply};
use redis::{Commands, RedisError};
use serde_json::Value;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum QueueError {
    #[error("Redis error: {0}")]
    Redis(#[from] RedisError),
    #[error("JSON error: {0}")]
    Json(#[from] serde_json::Error),
}

/// Abstract message queue interface.
pub trait MessageQueue: Send + Sync {
    fn publish(&self, topic: &str, message: &Value) -> Result<String, QueueError>;
    fn consume(&self, topic: &str, count: usize, block_ms: usize) -> Result<Vec<(String, Value)>, QueueError>;
    fn ack(&self, topic: &str, message_id: &str) -> Result<(), QueueError>;
    fn extend_visibility(&self, topic: &str, message_id: &str, seconds: u64) -> Result<(), QueueError>;
}

/// Redis Streams implementation with consumer groups.
pub struct RedisStreamQueue {
    client: redis::Client,
    consumer_group: String,
    consumer_name: String,
}

impl RedisStreamQueue {
    pub fn new(url: &str, consumer_group: &str) -> Result<Self, QueueError> {
        let client = redis::Client::open(url)?;
        let consumer_name = format!("worker-{}", &uuid::Uuid::new_v4().to_string()[..8]);
        Ok(Self {
            client,
            consumer_group: consumer_group.to_string(),
            consumer_name,
        })
    }

    fn ensure_group(&self, conn: &mut redis::Connection, topic: &str) -> Result<(), QueueError> {
        let result: Result<(), RedisError> =
            redis::cmd("XGROUP")
                .arg("CREATE")
                .arg(topic)
                .arg(&self.consumer_group)
                .arg("0")
                .arg("MKSTREAM")
                .query(conn);
        match result {
            Ok(()) => Ok(()),
            Err(e) if e.to_string().contains("BUSYGROUP") => Ok(()),
            Err(e) => Err(QueueError::Redis(e)),
        }
    }
}

impl MessageQueue for RedisStreamQueue {
    fn publish(&self, topic: &str, message: &Value) -> Result<String, QueueError> {
        let mut conn = self.client.get_connection()?;
        self.ensure_group(&mut conn, topic)?;
        let payload = serde_json::to_string(message)?;
        let id: String = conn.xadd(topic, "*", &[("data", &payload)])?;
        Ok(id)
    }

    fn consume(&self, topic: &str, count: usize, block_ms: usize) -> Result<Vec<(String, Value)>, QueueError> {
        let mut conn = self.client.get_connection()?;
        self.ensure_group(&mut conn, topic)?;

        let opts = StreamReadOptions::default()
            .group(&self.consumer_group, &self.consumer_name)
            .count(count)
            .block(block_ms);

        let reply: StreamReadReply = conn.xread_options(&[topic], &[">"], &opts)?;
        let mut messages = Vec::new();
        for key in reply.keys {
            for entry in key.ids {
                if let Some(redis::Value::BulkString(bytes)) = entry.map.get("data") {
                    let data_str = String::from_utf8_lossy(bytes);
                    let value: Value = serde_json::from_str(&data_str)?;
                    messages.push((entry.id, value));
                }
            }
        }
        Ok(messages)
    }

    fn ack(&self, topic: &str, message_id: &str) -> Result<(), QueueError> {
        let mut conn = self.client.get_connection()?;
        let _: () = conn.xack(topic, &self.consumer_group, &[message_id])?;
        Ok(())
    }

    fn extend_visibility(&self, _topic: &str, _message_id: &str, _seconds: u64) -> Result<(), QueueError> {
        // Redis Streams don't have visibility timeout — no-op.
        Ok(())
    }
}
